rust/src/sdk/api.rs

```diff
@@ -0,0 +1,448 @@
+//!    protocol work as well as the networking with the coordinator.
+//! 2. Optionally request status information:
+//!     - `is_next_round()` indicates if another round of the PET protocol has started.
+//!     - `is_update_participant()` indicates if this participant is eligible to submit a
+//!       trained local model.
+//! 3. Get the latest global model with `get_model_N()`, where `N` is the primitive data type.
+//!    Currently, `f32`, `f64`, `i32` and `i64` are supported. The function returns a mutable
+//!    slice to the primitive model, whereas the primitive model itself is cached within the
+//!    client.
+//! 4. Register a trained local model with `update_model_N()`, which takes either a slice to
+//!    the cached primitive model or a slice to a foreign memory location.
+//! 5. Destroy the client with `drop_client()`.
+
+use std::{
+    cmp::Ordering,
+    iter::{FromIterator, IntoIterator, Iterator},
+    mem,
+    os::raw::{c_double, c_float, c_int, c_long, c_ulong},
+    ptr,
+};
+
+use num::{bigint::BigInt, rational::Ratio, traits::Zero};
+
+use crate::{
+    mask::model::{FromPrimitives, IntoPrimitives, Model},
+    participant::{Participant, Task},
+};
+
+/// Generates a struct to hold the C equivalent of `&mut [N]` for a primitive data type `N` and
+/// implements a consuming iterator for the struct. The arguments `$prim_rust` and `$prim_c` are
+/// the corresponding Rust and C primitive data types.
+macro_rules! PrimModel {
+    ($prim_rust:ty, $prim_c:ty $(,)?) => {
+        paste::item! {
+            #[derive(Clone, Copy)]
+            #[repr(C)]
+            /// A model of primitive data type represented as a mutable slice which can be accessed
+            /// from C. It holds a raw pointer `ptr` to the array of primitive values and its length
+            /// `len`.
+            pub struct [<PrimitiveModel $prim_rust:upper>] {
+                pub ptr: *mut $prim_c,
+                pub len: c_ulong,
+            }
+
+            /// An iterator that moves out of a primitive model.
+            pub struct [<IntoIter $prim_rust:upper>] {
+                model: [<PrimitiveModel $prim_rust:upper>],
+                count: isize,
+            }
+
+            impl IntoIterator for [<PrimitiveModel $prim_rust:upper>] {
+                type Item = $prim_rust;
+                type IntoIter = [<IntoIter $prim_rust:upper>];
+
+                /// Creates an iterator from a primitive model.
+                fn into_iter(self) -> Self::IntoIter {
+                    Self::IntoIter {
+                        model: self,
+                        count: -1_isize,
+                    }
+                }
+            }
+
+            impl Iterator for [<IntoIter $prim_rust:upper>] {
+                type Item = $prim_rust;
+
+                /// Advances the iterator and returns the next primitive value. Returns `None` when
+                /// the iteration is finished.
+                ///
+                /// # Safety
```

Comment:
I don't think we can do that even if we document it unfortunately, because this breaks the unsafe contract: a function that can result in UB if fed with the wrong input must be marked as unsafe.

Reply:
we can't alter the trait's definition, but we can wrap this in a private module such that nobody can accidentally access this. then this is only called in functions which are marked as unsafe anyways.

Reply:
That sounds like a good idea to me :+1: